### ISSUE TITLE
False positive `unnecessary-dict-index-lookup` and `consider-using-dict-items` if value associated with key is changed in loop

### ISSUE THREAD

**Issue by akosthekiss** [PyCQA/pylint#4630]
### Steps to reproduce

Given a file `d.py`:
```python
"""d.py"""

d = {'key': 'value'}

print(d)

for k, _ in d.items():
    d[k] += 'VALUE'
    if 'V' in d[k]:  # if d[k] is replaced with _, the semantics change
        print('found V')

print(d)

for k in d:  # if this gets rewritten to d.items(), we are back to the above problem
    d[k] += '123'
    if '1' in d[k]:
        print('found 1')

print(d)
```

### Current behavior

Result of `pylint d.py`:
```
************* Module d
d.py:9:14: R1733: Unnecessary dictionary index lookup, use '_' instead (unnecessary-dict-index-lookup)
d.py:14:0: C0206: Consider iterating with .items() (consider-using-dict-items)
```

### Expected behavior

No diagnostics.

Following the recommendation for line 9 (i.e., rewriting `d[k]` to `_`) would incorrectly change the semantics of the program, as `d[k]` has already been changed, so it has a different value than `_`.

Following the recommendation for line 14 (i.e., rewriting `for k in d` to `for k, v in d.items()` would lead back to the above problem.

### pylint --version output

Result of `pylint --version` output:

```
pylint 3.0.0-a4
astroid 2.6.1
Python 3.6.9 (default, Jan 26 2021, 15:33:00) 
[GCC 8.4.0]
```

```
pylint 2.9.0
astroid 2.6.1
Python 3.6.9 (default, Jan 26 2021, 15:33:00) 
[GCC 8.4.0]
```

**Comment by cdce8p** [comment]
@yushao2 Would you like to take a look at it? It seems we need to ignore `unnecessary-dict-index-lookup` and `consider-using-dict-items` if an `Assign` or `AugAssign` is used.

**Comment by yushao2** [comment]
> @yushao2 Would you like to take a look at it? It seems we need to ignore `unnecessary-dict-index-lookup` and `consider-using-dict-items` if an `Assign` or `AugAssign` is used.

raised a pr for it: #4640
